### PR TITLE
Cosmos: Fix partition key range cache to use change-feed pagination

### LINKED ISSUE #26766: [Cosmos] azcosmos sends A-IM header with wrong case ("Incremental feed" vs canonical "Incremental Feed")
### Summary

The Go Cosmos SDK (`sdk/data/azcosmos`) sends the `A-IM` HTTP header with the value `"Incremental feed"` (lowercase **f**) on every ChangeFeed request. The wire-canonical / service-side constant is `"Incremental Feed"` (uppercase **F**), and that's also what every other Cosmos SDK and the service tests use.

### Where

- Constant: `sdk/data/azcosmos/cosmos_http_constants.go:97`
  ```go
  cosmosHeaderValuesChangeFeed string = "Incremental feed"
  ```
- Used unconditionally on every ChangeFeed call: `sdk/data/azcosmos/cosmos_change_feed_request_options.go:37`
  ```go
  headers[cosmosHeaderChangeFeed] = cosmosHeaderValuesChangeFeed
  ```

### Authoritative value

From the service:
```csharp
public static class A_IMHeaderValues
{
    public const string IncrementalFeed = "Incremental Feed";
    public const string FullFidelityFeed = "Full-Fidelity Feed";
    public const string IncrementalFullFidelityFeed = "Incremental Full-Fidelity Feed";
}
```

The .NET SDK sends the canonical value

### Why this is mostly harmless today, but should still be fixed

The compute gateway compares the header with `OrdinalIgnoreCase` (in `Product/Cosmos/RoutingGateway/Runtime/RequestHandlers/PartitionKeyRangesRequestHandler.cs`):
```csharp
bool isChangeFeed =
    request.Headers[HttpConstants.HttpHeaders.A_IM] != null &&
    request.Headers[HttpConstants.HttpHeaders.A_IM]
        .Equals(HttpConstants.A_IMHeaderValues.IncrementalFeed, StringComparison.OrdinalIgnoreCase);
```
so functionality is unaffected. **But:**

1. Anything downstream that does case-sensitive comparison (raw network captures, third-party HTTP middleware, future backend or gateway code paths) will silently mis-classify Go SDK ChangeFeed requests as ReadFeed.
2. It diverges from every other Cosmos SDK on the wire — makes cross-SDK telemetry harder to reason about.
3. The header value is part of the documented HTTP contract; the SDK should match the spec exactly.

### Suggested fix

```diff
- cosmosHeaderValuesChangeFeed    string = "Incremental feed"
+ cosmosHeaderValuesChangeFeed    string = "Incremental Feed"
```
plus add a unit test that asserts the literal byte sequence `"Incremental Feed"` (not just equality to the constant), so a future typo in the constant gets caught.

### Discovery context

Surfaced while running the Go SDK perf tool (#26764) at sustained ~70K ChangeFeed ops/min/pod and grepping gateway text-trace logs for `requestType ChangeFeed` rate to validate that the Go SDK was actually being recognized as ChangeFeed — which it is today thanks to the case-insensitive compare, but only by luck.

### PR DESCRIPTION
**Pull request by simorenoh**
Summary

Fixes the partition key range cache to use the change-feed mechanism with pagination for both full and incremental refreshes, matching the behavior of the .NET and Python Cosmos SDKs. This resolves a customer-reported issue where ReadMany fails with incomplete partition key range set after full refresh on containers with many partitions.

Root Cause

The previous implementation did a single plain GET request to fetch partition key ranges without pagination or the change-feed mechanism. On containers with many partitions where the response exceeds a single page, this resulted in an incomplete set of ranges. The validation added with the PK range cache surfaced this latent issue.

Changes

 - Always use change-feed headers: Sets A-IM: Incremental feed and x-ms-max-item-count: -1 on all PK range fetch requests
 - Full refresh pagination: Loops fetching change-feed pages until 304 Not Modified, accumulating all ranges before building the routing map
 - Incremental refresh accumulation: Accumulates ALL incremental change-feed pages before calling tryCombine once (rather than per-page). This correctly handles cascading splits (A→B+C on page 1, B→D+E on page 2) that would otherwise cause unnecessary full-refresh fallbacks
 - Improved error diagnostics: Error messages now explicitly indicate cache refresh failure, include range count, and describe the specific gap detected (e.g., gap between range id=3 (max=X) and range id=5 (min=Y))

Testing

 - Updated existing tests to account for 304 termination responses
 - Added tests for multi-page full refresh, split-during-fetch (parent filtering), header verification, and cascading splits across incremental pages

### CODE REVIEW

File: sdk/data/azcosmos/cosmos_partition_key_range_cache_test.go [hunk 1]
```diff
@@ -377,3 +388,261 @@ func Test_partitionKeyRangeCache_incrementalRefresh_contextCancelled(t *testing.
...
+
+	// Second request: should have A-IM, max-item-count, AND If-None-Match with the ETag from page 1
+	secondReq := capture.requests[1]
+	require.Equal(t, cosmosHeaderValuesChangeFeed, secondReq.Header.Get(cosmosHeaderChangeFeed))
+	require.Equal(t, cosmosHeaderValuesMaxItemAll, secondReq.Header.Get(cosmosHeaderMaxItemCount))
+	require.Equal(t, "etag1", secondReq.Header.Get(headerIfNoneMatch))
+}
+
+func Test_partitionKeyRangeCache_incrementalRefresh_cascadingSplitAcrossPages(t *testing.T) {
+	// Scenario: Page 1 returns only ONE child of a split (range "1"), making the
```
Comment (FabianMeiswinkel) [inline sdk/data/azcosmos/cosmos_partition_key_range_cache_test.go:704]: Please add tests cases also injecting empty pages - returning 200 with no changes - to ensure we keep draining until we get 304 Not Modified

**Review by FabianMeiswinkel — CHANGES_REQUESTED** [review]
LGTM except for one missing Test case

File: sdk/data/azcosmos/cosmos_partition_key_range_cache.go [hunk 2]
```diff
@@ -120,14 +120,57 @@ func (c *partitionKeyRangeCache) invalidate(containerRID string) {
...
+	for i := 0; i < maxChangeFeedIterations; i++ {
+		if err := ctx.Err(); err != nil {
+			return changeFeedResult{}, err
+		}
+		ranges, newETag, err := fetchPartitionKeyRanges(ctx, containerLink, currentETag, client)
+		if err != nil {
+			return changeFeedResult{}, err
+		}
+
+		if len(ranges) == 0 {
```
Comment (xinlian12) [inline sdk/data/azcosmos/cosmos_partition_key_range_cache.go]: can we reliably say only 304 will result in len(ranges) == 0? for change feed, there can be cases where response is empty with 200

File: sdk/data/azcosmos/cosmos_partition_key_range_cache.go [hunk 3]
```diff
@@ -120,14 +120,57 @@ func (c *partitionKeyRangeCache) invalidate(containerRID string) {
...
+			}
+			return changeFeedResult{ranges: allRanges, finalETag: currentETag, completed: true}, nil
+		}
+		allRanges = append(allRanges, ranges...)
+		if newETag != "" {
+			currentETag = newETag
+		}
+	}
+	// Loop cap reached without 304
+	return changeFeedResult{ranges: allRanges, finalETag: currentETag, completed: false}, nil
```
Comment (xinlian12) [inline sdk/data/azcosmos/cosmos_partition_key_range_cache.go:170]: can we also add logs to indicate sdk exit without reaching 304?

**Review by xinlian12 — APPROVED** [review]
LGTM, thanks

**Comment by xinlian12** [comment]
@sdkReviewAgent

**Review by kushagraThapar — APPROVED** [review]
LGTM — thank you for the careful split between full-refresh pagination, incremental accumulation, and the cascading-splits-across-pages handling. A few things I verified while looking through this:

- **Fabian's empty-pages ask is fully addressed in `8beab079`** — `Test_partitionKeyRangeCache_fullRefresh_emptyPagesBeforeData` and `..._incrementalRefresh_emptyPagesBeforeData` exercise the 200-empty → 200-empty → 200-data → 304 sequence on both refresh paths, and the supporting `notModified bool` refactor on `fetchPartitionKeyRangesResult` cleanly removes the prior `len(ranges) == 0` ambiguity.
- **Cross-SDK parity checks out** — the change-feed mechanism (A-IM + max-item-count=-1), 304-as-terminator, and accumulate-then-`tryCombine`-once all align with .NET, Python, Java, and Rust. The cascading-splits test (`..._cascadingSplitAcrossPages`) is a really nice piece of evidence for why the per-page → batch-then-combine refactor is the right call.
- **The new diagnostics are genuinely nicer** — having the gap description (range count + which two ranges have the boundary mismatch) in the error string will be a meaningful upgrade for future SEV investigations.

A couple of small, optional things I noticed (none required for this PR — happy to file as follow-up issues if you'd like):

1. xinlian12's open ask about an `azlog` line at the cap-exhaustion path inside `fetchAllChangeFeedPages` (and the incremental fall-through at line 216) is still open — would help close his thread, and the package already uses `azlog`/`log.Write` elsewhere (`cosmos_client.go:117/127/560`).
2. There's no test that drives the loop past `maxChangeFeedIterations = 1000` — could be worth pinning both the cap behavior and the diagnostic string with a tiny test that lowers the cap or returns 200-with-data forever.
3. The doc comment at `cosmos_partition_key_range_cache.go:247-249` says the headers match ".NET and Python", but after the casing flip in `50f01ddf` the literal `"Incremental Feed"` actually matches Java/Rust (Python uses lowercase 'f'). Service is case-insensitive, so wire-behavior is unchanged — purely a reader-clarity thing.

None of these block — approving on the strength of Fabian's ask being addressed and the cross-SDK parity holding. Thanks Simon!

File: sdk/data/azcosmos/cosmos_partition_key_range_cache.go (hunk 2 above)
Comment (simorenoh) [reply sdk/data/azcosmos/cosmos_partition_key_range_cache.go]: added tests for this :)

File: sdk/data/azcosmos/cosmos_partition_key_range_cache.go (hunk 3 above)
Comment (simorenoh) [reply sdk/data/azcosmos/cosmos_partition_key_range_cache.go:170]: added

File: sdk/data/azcosmos/cosmos_partition_key_range_cache_test.go (hunk 1 above)
Comment (simorenoh) [reply sdk/data/azcosmos/cosmos_partition_key_range_cache_test.go:704]: added

File: sdk/data/azcosmos/cosmos_partition_key_range_cache_test.go [hunk 4]
```diff
@@ -318,6 +324,11 @@ func Test_partitionKeyRangeCache_incrementalRefresh_mergeFailure_fullRefresh(t *
 		mock.WithHeader(cosmosHeaderEtag, "etag3"),
 		mock.WithStatusCode(200),
 	)
+	// 304 Not Modified — terminates the full change-feed refresh loop
```
Comment (xinlian12) [inline sdk/data/azcosmos/cosmos_partition_key_range_cache_test.go:333]: 🔴 **Blocking — Test Correctness: Merge-failure test no longer exercises the full-refresh fallback**

This test claims to validate "tryCombine returns nil → falls through to full refresh," but with the new `fetchAllChangeFeedPages` pagination, it silently passes via the **incremental success path** instead.

**Trace:** The mock server queues responses FIFO. When the incremental `fetchAllChangeFeedPages` runs, it loops until 304 — consuming all three responses in one call:
1. Response 1 (200): 1 child range `{ID:"1", parents:["0"]}` — accumulated
2. Response 2 (200): 3 complete ranges `{ID:"1"}, {ID:"2"}, {ID:"3"}` — accumulated  
3. Response 3 (304): terminates the loop

`tryCombine` then receives all 4 accumulated ranges. Parent "0" is filtered via `combinedGone`. Duplicate ID "1" deduplicates in the map. Result: 3 ranges forming a complete covering → **tryCombine succeeds**. The full-refresh code path at line 220 (`fetchAllChangeFeedPages(ctx, containerLink, "", client)`) is **never reached**.

**Why it matters:** The incremental-merge-failure → full-refresh fallback is a critical reliability path. If a regression broke it, this test wouldn't catch it.

**Fix:** Insert a 304 between the incremental and full-refresh responses to properly partition the two code paths:

```go
// Incremental response: partial child leaving a gap
srv.AppendResponse(/* 200, 1 child with parent "0" */)
// 304 terminates incremental change-feed loop with incomplete data
srv.AppendResponse(mock.WithStatusCode(304), mock.WithHeader(cosmosHeaderEtag, "etag2"))
// Full refresh: complete set of 3 ranges
srv.AppendResponse(/* 200, 3 complete ranges */)
// 304 terminates full refresh loop
srv.AppendResponse(mock.WithStatusCode(304), mock.WithHeader(cosmosHeaderEtag, "etag3"))

```

This ensures `fetchAllChangeFeedPages` on the incremental path returns only the 1 child range, `tryCombine` sees the gap and returns nil, and the code falls through to the full refresh — which is what the test name promises to verify.

⚠️ AI-generated review — may be incorrect. Agree? → resolve the conversation. Disagree? → reply with your reasoning.

File: sdk/data/azcosmos/cosmos_http_constants.go [hunk 5]
```diff
@@ -96,7 +96,8 @@ const (
 const (
 	cosmosHeaderValuesPreferMinimal string = "return=minimal"
 	cosmosHeaderValuesQuery         string = "application/query+json"
-	cosmosHeaderValuesChangeFeed    string = "Incremental feed"
+	cosmosHeaderValuesChangeFeed    string = "Incremental Feed"
```
Comment (xinlian12) [inline sdk/data/azcosmos/cosmos_http_constants.go:99]: 🟡 **Recommendation — Cross-SDK: Header casing change affects all change-feed operations**

This changes `cosmosHeaderValuesChangeFeed` from `"Incremental feed"` (lowercase f) to `"Incremental Feed"` (capital F). The blast radius extends **beyond PK range fetching** — this constant is also used in `cosmos_change_feed_request_options.go:45` for all user-facing change-feed operations (`GetChangeFeed`).

Cross-SDK comparison:

| SDK | A-IM Value |
|-----|-----------|

| REST API docs | `"Incremental feed"` (lowercase f) |

| Python | `"Incremental feed"` (lowercase f) |

| Java | `"Incremental Feed"` (uppercase F) |

| Rust | `"Incremental Feed"` (uppercase F) |

| .NET | Private constant — unknown |

| **Go (pre-PR)** | `"Incremental feed"` ✅ matched spec |

| **Go (this PR)** | `"Incremental Feed"` ← changed |

The [REST API documentation](https://learn.microsoft.com/en-us/rest/api/cosmos-db/list-documents) explicitly says: *"Must be set to `Incremental feed`"* (lowercase f). The Cosmos service almost certainly accepts both, but:
1. The original value was spec-correct
2. All tests compare against the constant reference, providing **no safety net** for casing regressions
3. The change is incidental to the PR's purpose (fixing pagination)

Consider reverting to `"Incremental feed"` to match the REST API spec and Python SDK, or confirming with the Cosmos service team that `"Incremental Feed"` is the intended canonical value.

⚠️ AI-generated review — may be incorrect. Agree? → resolve the conversation. Disagree? → reply with your reasoning.

File: sdk/data/azcosmos/cosmos_partition_key_range_cache.go [hunk 6]
```diff
@@ -138,79 +181,87 @@ func (c *partitionKeyRangeCache) refreshEntry(
...
-				return currentMap, nil
+				return entry.routingMap, nil
 			}
 
-			merged := currentMap.tryCombine(ranges, newETag)
-			if merged == nil {
-				// Incremental merge failed — fall through to full refresh
-				break
+			merged := previousMap.tryCombine(result.ranges, result.finalETag)
+			if merged != nil && isCompleteSetOfRanges(merged.orderedRanges) {
```
Comment (xinlian12) [inline sdk/data/azcosmos/cosmos_partition_key_range_cache.go]: 🟢 **Suggestion — Redundant completeness check after tryCombine**

`isCompleteSetOfRanges(merged.orderedRanges)` is redundant here. `tryCombine` already calls `isCompleteSetOfRanges` internally (line 103 of `cosmos_collection_routing_map.go`) and returns `nil` if the ranges are incomplete. When `merged != nil`, completeness is guaranteed by `tryCombine`'s contract.

The check is harmless but could mislead future developers into thinking `tryCombine` might return a non-nil map with incomplete ranges. Consider either:
- Removing `&& isCompleteSetOfRanges(merged.orderedRanges)` and trusting `tryCombine`'s contract
- Adding a clarifying comment: `// Defensive: tryCombine already validates completeness internally`

⚠️ AI-generated review — may be incorrect. Agree? → resolve the conversation. Disagree? → reply with your reasoning.

**Comment by xinlian12** [comment]
✅ **Review complete** (53:25)

Posted **3** inline comment(s).

<sub>Steps: ✓ context, correctness, cross-sdk, design, history, past-prs, synthesis, test-coverage</sub>

File: sdk/data/azcosmos/cosmos_partition_key_range_cache.go (hunk 6 above)
Comment (simorenoh) [reply sdk/data/azcosmos/cosmos_partition_key_range_cache.go]: done

File: sdk/data/azcosmos/cosmos_partition_key_range_cache_test.go (hunk 4 above)
Comment (simorenoh) [reply sdk/data/azcosmos/cosmos_partition_key_range_cache_test.go:333]: done

**Review by xinlian12 — APPROVED** [review]
LGTM, thanks

**Review by FabianMeiswinkel — APPROVED** [review]
LGTM
